src/plugins/intel_gpu/src/graph/impls/ocl/assign.cpp

```diff
@@ -0,0 +1,65 @@
+// Copyright (C) 2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "assign_inst.h"
+#include "primitive_base.hpp"
+#include "impls/implementation_map.hpp"
+
+namespace cldnn {
+namespace ocl {
+
+struct assign_impl : public typed_primitive_impl_ocl<assign> {
+    using parent = typed_primitive_impl_ocl<assign>;
+    using parent::parent;
+
+    std::unique_ptr<primitive_impl> clone() const override {
+        return make_unique<assign_impl>(*this);
+    }
+
+    event::ptr execute_impl(const std::vector<event::ptr>& events, assign_inst& instance) override {
+        const auto arg = instance.argument;
+        const auto variable_id = arg.variable_id;
+
+        auto& variables = instance.get_network().get_variables();
+
+        auto var_it = variables.find(variable_id);
+        if (var_it == variables.end()) {
+            CLDNN_ERROR_MESSAGE(instance.id(), "Variable " + variable_id + " not found");
+        }
+
+        auto& variable = var_it->second;
+
+        if (variable.memory->get_layout() != arg.output_layout) {
+            CLDNN_ERROR_MESSAGE(instance.id(), "Layout mismatch");
+        }
+
+        std::vector<event::ptr> tmp_events(events);
+        auto& stream = instance.get_network().get_stream();
+
+        stream.enqueue_barrier(); // input must be ready before proceed
+
+        const auto ev_set_memory = variable.memory->copy_from(stream, instance.input_memory());
+        tmp_events.push_back(ev_set_memory);
+        variable.is_set = true;
+
+        const auto ev_set_output = instance.output_memory().copy_from(stream, *variable.memory);
```

Comment:
Do I understand correctly that this double copy is needed since variable.memory has different buffer with output_memory() and if they were the same, then we'd need single copy only?

If yes, then consider customizing read_value_inst c-tor:

```cpp
// additional false in parent c-tor disables default output alloc logic
read_value_inst::typed_primitive_inst(network& network, const read_value_node& node) : parent(network, node, false) {
    auto& variables = instance.get_network().get_variables();

    auto var_it = variables.find(variable_id);
    if (var_it == variables.end()) {
         CLDNN_ERROR_MESSAGE(instance.id(), "Variable " + variable_id + " not found");
    }
    auto& variable = var_it->second;
    _output = variable.memory;
} 

```

Reply:
Done.